File: AM_Program/ISpectraFileProcessor.vb
'*********************************************************************************************************
' Written by Dave Clark for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Copyright 2006, Battelle Memorial Institute
' Created 06/07/2006
'
' Last modified 07/29/2008
'*********************************************************************************************************

Imports System.Collections.Specialized

Namespace AnalysisManagerBase

	Public Interface ISpectraFileProcessor
		'*********************************************************************************************************
		'Defines minimum required functionality for classes that will generate spectra files
		'*********************************************************************************************************

#Region "Enums"
		Enum ProcessResults
			'Return values for MakeDTA and Abort functions
			SF_SUCCESS = 0		'Operation succeeded
			SF_FAILURE = -1		'Operation failed
			SF_NO_FILES_CREATED = -2		'Spectra file creation didn't fail, but no output files were created
			SF_ABORTED = -3		'Spectra file creation aborted
		End Enum

		Enum ProcessStatus
			'Return value for status property
			SF_STARTING		'Plugin initialization in progress
			SF_RUNNING		'Plugin is attempting to do its job
			SF_COMPLETE		'Plugin successfully completed its job
			SF_ERROR		'There was an error somewhere
			SF_ABORTING		'An ABORT command has been received; plugin shutdown in progress
		End Enum
#End Region

#Region "Structures"
		Structure InitializationParams
			'Dim SourceFolderPath As String
			'Dim OutputFolderPath As String
			'Dim MiscParams As StringDictionary
			Dim DebugLevel As Integer
            Dim MgrParams As IMgrParams
			Dim JobParams As IJobParams
			Dim StatusTools As IStatusFile
		End Structure
#End Region

#Region "Properties"
		'		Property SourceFolderPath() As String	' (in)  – path to folder containing the raw spectra file

		'		Property OutputFolderPath() As String	' (in) – path to folder where generated DTAs are to be placed

		'		WriteOnly Property MiscParams() As StringDictionary	'For passing miscelleneous parameters (not presently used)

		ReadOnly Property Status() As ISpectraFileProcessor.ProcessStatus	'Allows calling program to get current status

		ReadOnly Property Results() As ISpectraFileProcessor.ProcessResults	 'Allows calling program to determine if DTA creation succeeded

		ReadOnly Property ErrMsg() As String	 'Error message describing any errors encountered

		Property DebugLevel() As Integer	'Allows control of debug information verbosity; 0=minimum, 5=maximum verbosity

        ReadOnly Property SpectraFileCount() As Integer 'Count of spectra files that have been created

		WriteOnly Property MgrParams() As IMgrParams	'Machine-specific parameters, such as file locations

		WriteOnly Property JobParams() As IJobParams	'Job-specific parameters

		WriteOnly Property StatusTools() As IStatusFile	 'Interface for updating task status
#End Region

#Region "Methods"
		Sub Setup(ByVal InitParams As InitializationParams)	 'Initializes parameters. Must be called before executing Start()

		Function Start() As ISpectraFileProcessor.ProcessStatus	 'Starts the spectra file creation process

		Function Abort() As ISpectraFileProcessor.ProcessStatus	 'Aborts spectra file creation
#End Region

	End Interface

End Namespace

File: AM_Shared/DataFileTools/ISpectraFileProcessor.cs

//*********************************************************************************************************
// Written by Dave Clark and Matthew Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Copyright 2006, Battelle Memorial Institute
// Created 06/07/2006
//
//*********************************************************************************************************

using AnalysisManagerBase.AnalysisTool;
using AnalysisManagerBase.JobConfig;
using AnalysisManagerBase.StatusReporting;

namespace AnalysisManagerBase.DataFileTools
{
    /// <summary>
    /// Initialization parameters for classes that implement ISpectraFileProcessor
    /// </summary>
    public struct SpectraFileProcessorParams
    {
        /// <summary>
        /// Debug level
        /// </summary>
        public int DebugLevel;

        /// <summary>
        /// Manager parameters
        /// </summary>
        public IMgrParams MgrParams;

        /// <summary>
        /// Job parameters
        /// </summary>
        public IJobParams JobParams;

        /// <summary>
        /// Status tools
        /// </summary>
        public IStatusFile StatusTools;

        /// <summary>
        /// Working directory
        /// </summary>
        public string WorkDir;

        /// <summary>
        /// Dataset name
        /// </summary>
        public string DatasetName;
    }

    /// <summary>
    /// Defines minimum required functionality for classes that will generate spectra files
    /// </summary>
    public interface ISpectraFileProcessor
    {
        /// <summary>
        /// Allows calling program to get current status
        /// </summary>
        ProcessStatus Status { get; }

        /// <summary>
        /// Allows calling program to determine if DTA creation succeeded
        /// </summary>
        ProcessResults Results { get; }

        /// <summary>
        /// Error message describing any errors encountered
        /// </summary>
        string ErrMsg { get; }

        /// <summary>
        /// Allows control of debug information verbosity; 0=minimum, 5=maximum verbosity
        /// </summary>
        int DebugLevel { get; set; }

        /// <summary>
        /// Path to the program used to create .DTA files
        /// </summary>
        string DtaToolNameLoc { get; }

        /// <summary>
        /// Count of spectra files that have been created
        /// </summary>
        int SpectraFileCount { get; }

        /// <summary>
        /// Percent complete (Value between 0 and 100)
        /// </summary>
        float Progress { get; }

        /// <summary>
        /// Machine-specific parameters, such as file locations
        /// </summary>
        IMgrParams MgrParams { set; }

        /// <summary>
        /// Job-specific parameters
        /// </summary>
        IJobParams JobParams { set; }

        /// <summary>
        /// Interface for updating task status
        /// </summary>
        IStatusFile StatusTools { set; }

        /// <summary>
        /// Initializes parameters. Must be called before executing Start()
        /// </summary>
        /// <param name="InitParams"></param>
        /// <param name="toolRunner"></param>
        void Setup(SpectraFileProcessorParams InitParams, AnalysisToolRunnerBase toolRunner);

        /// <summary>
        /// Starts the spectra file creation process
        /// </summary>
        ProcessStatus Start();

        /// <summary>
        /// Aborts spectra file creation
        /// </summary>
        ProcessStatus Abort();
    }
}